code/datums/actions/actions.dm

```diff
@@ -184,93 +184,125 @@
-	var/spot
-	New(var/obj/item/sheet/csheet, var/cobjtype, var/ccost, var/datum/material/cmat, var/camount, var/cicon, var/cicon_state, var/cobjname, var/post_action_callback = null, var/obj/item/sheet/csheet2, var/ccost2, var/spot)
-		..()
-		icon = cicon
-		icon_state = cicon_state
-		sheet = csheet
-		objtype = cobjtype
-		cost = ccost
-		mat = cmat
-		amount = camount
-		objname = cobjname
-		callback = post_action_callback
-		src.spot = spot
-		if (csheet2)
-			sheet2 = csheet2
-		if (ccost2)
-			cost2 = ccost2
+	/// Custom name of the object for messages if needed, otherwise just the object's initial name
+	var/obj_name
+	var/obj_turf
+	var/obj_type
+	/// Amount of the object which will be made when we're done
+	var/obj_amt
+	var/obj/item/sheet/sheet1
+	var/obj/item/sheet/sheet2
+	var/cost1
+	var/cost2
+	/// Are we using a second sheet (e.g. Glass table parts)
+	var/has_sheet2
+	var/datum/material/obj_mat
+	/// Callback once the thing is constructed
+	var/post_callback
+	//New(var/obj/item/sheet/csheet, var/obj/cobjtype, var/ccost, var/datum/material/cmat, var/camount, var/cicon, var/cicon_state, var/cobjname, var/post_action_callback = null, var/obj/item/sheet/csheet2, var/ccost2, var/spot)
+	New(var/obj/otype, var/target, var/oamt, var/btime, var/s1, var/c1, var/s2, var/c2, var/datum/material/omat, var/c_icon = null, var/c_icon_state = null, var/callback = null, var/name = null)
+		..()
+		resumable = FALSE
+		obj_type = otype
+		obj_name = (!name) ? initial(otype.name) : name
+		obj_turf = get_turf(target || owner)
+		obj_amt = oamt
+		obj_mat = omat
+		duration = btime
+		sheet1 = s1
+		cost1 = c1
+		sheet2 = s2
+		cost2 = c2
+		has_sheet2 = (s2 != null)
+		post_callback = callback
+		// You need both to set a custom icon, there's no warning but you are expected to know this
+		if (c_icon && c_icon_state)
+			icon = c_icon
+			icon_state = c_icon_state
+		else
+			icon = initial(otype.icon)
+			icon_state = initial(otype.icon_state)
+
+	/// Return TRUE if both sheets are there and valid, else false
+	proc/has_valid_sheets()
+		if (QDELETED(sheet1) || (has_sheet2 && QDELETED(sheet2)))
+			boutput(owner, "<span class='notice'>You have nothing to build with!</span>")
+			return FALSE
+		if (sheet1.amount < cost1)
+			boutput(owner, "<span class='notice'>You don't have enough [sheet1]\s to build \the [obj_name] with!</span>")
+			return FALSE
+		if (has_sheet2 && sheet2.amount < cost2)
+			boutput(owner, "<span class='notice'>You don't have enough [sheet2]\s to build \the [obj_name] with!</span>")
+			return FALSE
+		if (ismob(owner))
+			var/mob/M = owner
+			if (!in_interact_range(sheet1, M))
+				boutput(owner, "<span class='notice'>You dropped \the [sheet1]\s, how are you going to finish \the [obj_name]?</span>")
+				return FALSE
+			if (has_sheet2 && !in_interact_range(sheet2, M))
+				boutput(owner, "<span class='notice'>\the [sheet2]\s have to be closer to build \the [obj_name]!</span>")
+				return FALSE
+		return TRUE
```

Comment:
Proc created here with the intent to eliminate repetition and increase clarity in the below Start/Update/End procs